controller/persistence/eventual_eventer.go

```diff
@@ -0,0 +1,607 @@
+
+func (a *EventualEventerBbolt) Trigger() (<-chan struct{}, error) {
+	if a.running.Get() {
+		doneNotify := make(chan struct{})
+		a.waiters.Store(cuid.New(), doneNotify)
+
+		select {
+		case a.trigger <- struct{}{}:
+		default:
+			//channel full, already queued for processing
+		}
+
+		return doneNotify, nil
+	}
+
+	return nil, errors.New("triggering is impossible, eventual eventer not started")
+}
+
+// deleteEventualEvent removes an eventual event by id from the bbolt backend store.
+func (a *EventualEventerBbolt) deleteEventualEvent(id string) error {
+	err := a.dbProvider.GetDb().Update(func(tx *bbolt.Tx) error {
+		ctx := boltz.NewMutateContext(tx)
+		return a.store.DeleteById(ctx, id)
+	})
+
+	if err == nil {
+		total := atomic.AddInt64(a.outstandingEvents, -1)
+		a.Emit(EventualEventRemovedName, &EventualEventRemoved{
+			Id:    id,
+			Total: total,
+		})
+	}
+
+	return err
+}
+
+// notifyWaiters closes all signaling channels returned in calls
+// to Trigger().
+func (a *EventualEventerBbolt) notifyWaiters() {
+	var waiterIds []string
+
+	a.waiters.Range(func(key, value interface{}) bool {
+		id := key.(string)
+		waiterChan := value.(chan struct{})
+		if waiterChan != nil {
+			close(waiterChan)
+		}
+
+		waiterIds = append(waiterIds, id)
+		return true
+	})
+
+	for _, id := range waiterIds {
+		a.waiters.Delete(id)
+	}
+}
+
+// getEventualEvents returns eventual events up from the persistent bbolt store. The number
+// returned is determined by batchSize and the order of the events are based on key sorting
+func (a *EventualEventerBbolt) getEventualEvents() ([]string, []*EventualEvent, error) {
+	var ids []string
+	var eventualEvents []*EventualEvent
+	err := a.dbProvider.GetDb().View(func(tx *bbolt.Tx) error {
+		var err error
+		ids, _, err = a.store.QueryIds(tx, fmt.Sprintf("limit %d", a.batchSize))
+
+		if err != nil {
+			return err
+		}
+
+		for _, id := range ids {
+			event, err := a.store.LoadOneById(tx, id)
+
+			if err != nil {
+				pfxlog.Logger().WithField("id", id).WithError(err).Errorf("error could not load event id %s", id)
+			} else {
+				eventualEvents = append(eventualEvents, event)
+			}
+		}
+		return nil
+	})
+
+	return ids, eventualEvents, err
+}
+
+// process is the main execution look for dealing with batches of eventual events, triggering listeners for
+// the eventual events and emitting normal processing events
+func (a *EventualEventerBbolt) process() {
+	processId := cuid.New()
+	totalBatches := int64(0)
+	totalEvents := int64(0)
+	totalListenersExecuted := int64(0)
+	totalEventIndex := int64(0)
+
+	a.Emit(EventualEventProcessingStartName, &EventualEventProcessingStart{
+		Id: processId,
+	})
+
+	defer func() {
+		a.notifyWaiters()
+
+		a.Emit(EventualEventProcessingDoneName, &EventualEventProcessingDone{
+			Id:                     processId,
+			TotalBatches:           totalBatches,
+			TotalEvents:            totalEvents,
+			TotalListenersExecuted: totalListenersExecuted,
+		})
+	}()
+
+	for {
```

Comment:
lots of code in the while loop here makes it hard to see where it exits.

Reply:
Split up just for you.